### PR TITLE
dnp3: fixed wrong flow direction identification - v2

### PR DESCRIPTION
**Pull request by ilya-bakhtin**
dnp3 is a communication between so-called master and outstation
in our terms master is a client and outstation is a server
DIR flag in dnp3 header is nonzero when a packet is from master
so if DIR is nonzero then packet is 'toserver'

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3772

Describe changes:

The condition when the direction is reversed is changed according to dnp3 specification.
The value of returned rdir is changed to follow the pattern used in the caller.

In a case of DNP3 banner direction may only be to client.

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

suricata-verify-pr: 310
#suricata-verify-repo:
#suricata-verify-branch:
#suricata-update-pr:
#suricata-update-repo:
#suricata-update-branch:
#libhtp-pr:
#libhtp-repo:
#libhtp-branch:

### CODE REVIEW

File: src/app-layer-dnp3.c [hunk 1]
```diff
@@ -271,20 +271,24 @@ static uint16_t DNP3ProbingParser(Flow *f, uint8_t direction,
         const uint8_t *input, uint32_t len,
         uint8_t *rdir)
 {
-    DNP3LinkHeader *hdr = (DNP3LinkHeader *)input;
+    const DNP3LinkHeader* const hdr = (const DNP3LinkHeader*)input;
```
Comment (catenacyber) [inline src/app-layer-dnp3.c:274]: Why do you use `const` ? to make sure we do not overwrite the input data ?

File: src/app-layer-dnp3.c [hunk 2]
```diff
@@ -271,20 +271,24 @@ static uint16_t DNP3ProbingParser(Flow *f, uint8_t direction,
...
+        return ALPROTO_DNP3;
+    }
 
     /* Check that we have the minimum amount of bytes. */
     if (len < sizeof(DNP3LinkHeader)) {
         SCLogDebug("Length too small to be a DNP3 header.");
         return ALPROTO_UNKNOWN;
     }
 
-    /* May be a banner. */
```
Comment (catenacyber) [inline src/app-layer-dnp3.c:282]: Why did you move this above the length check ?

**Comment by catenacyber** [comment]
I edited the first comment, so that CI runs with the suricata-verify patch

**Comment by ilya-bakhtin** [comment]
parameter input is a const pointer. it's not a good style to cast it to non-const without reason.

banner does not assume a header so it may be shorter than the header length.

**Review by catenacyber — APPROVED** [review]
Looks good to me.

Travis failure seems to be due ti the fact that it does not use the suricata-verify PR

**Comment by catenacyber** [comment]
Merged in commit 12dfc33e05d042447027fa446fd07babd8086081
